config/base.py

```diff
@@ -0,0 +1,144 @@
+"""Base configuration holder."""
+
+from __future__ import annotations
+from typing import TYPE_CHECKING, List, Optional, TypeVar, Callable, Union
+
+if TYPE_CHECKING:
+    # Avoid cyclic imports used for typing only.
+    from database.tables import ConfigsTable
+
+
+T = TypeVar("T")
+
+
+class ConfigVariable():
+    """Wraps config property to implement lazy loading and automatic saving.
+
+    This class is an implementation close to Python's property() function. It
+    provides a user friendly interface for configuration function, documenting
+    variables per configuration, providing default constructor, list of allowed
+    values etc.
+
+    Usage:
+    >>> class MyConfig(Config):
+    ...     my_variable = ConfigVariable(
+    ...         key = "my_variable",             # Unique identifier
+    ...         name = "My awesome variable",    # User friendly name
+    ...         description = "Lots of things",  # Opt: documentation
+    ...         values = ["foo", "bar"],         # Opt: allowed values
+    ...         constructor = str,               # Opt: constructor
+    ...         default = "foo")                 # Opt: default value.
+    ...
+    >>> config = MyConfig()
+    >>> print(config.my_variable)
+    foo
+    >>> config.my_variable = "bar"
+    >>> print(config.my_variable)
+    bar
+
+    :arg key: a unique identifier for this key, used to key the database
+              entries. Ideally, you want to make it match the property's name
+              used when declaring the configuration implementation.
+    :arg name: a short, user friendly name, used to document the variable.
+    :arg description: a long description, used for documentation.
+    :arg values: optional list of possibly values. If specified, any other
+                 values set to the variable will result in an exception.
+    :arg constructor: optional constructor, used when accessing the variable.
+    :arg default: optional default value for this property.
+    """
+
+    def __init__(self, *, key: str, name: str, description: str = "",
+                 values: Optional[List[str]] = None,
+                 constructor: Optional[Callable[[str], T]] = None,
+                 default: Optional[str] = None):
+        self.key = key
+        self.name = name
+        self.description = description
+        self.values = values
+        self.constructor = constructor
+        self.default = default
+
+    def __set__(self, config: BaseConfig, value: str):
+        """Calls BaseConfig._set on the variable, saving it in the database."""
+        if self.values is not None and value not in self.values:
```

Comment:
`if self.values is not None` -> `if self.values`

Reply:
Also, I don't have strong feelings about this kind of syntaxes, and I would overall not mind leaving it as is, but I just wanted to open the discussion on whether or not you have a specific reason for using this syntax